miner/currency_blocklist.go

```diff
@@ -0,0 +1,136 @@
+package miner
+
+import (
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+)
+
+const hours uint64 = 60 * 60
+
+var evictionTimeoutSeconds uint64 = 2 * hours
+
+type AddressBlocklist struct {
+	mux        *sync.RWMutex
+	currencies map[common.Address]*types.Header
+	// fee-currencies blocked at headers with an older timestamp
+	// will get evicted when evict() is called
+	headerEvictionTimeoutSeconds uint64
+	oldestHeader                 *types.Header
+}
+
+func NewAddressBlocklist() *AddressBlocklist {
+	return &AddressBlocklist{
+		mux:                          &sync.RWMutex{},
+		currencies:                   map[common.Address]*types.Header{},
+		headerEvictionTimeoutSeconds: evictionTimeoutSeconds,
+		oldestHeader:                 nil,
+	}
+}
+
+func (b *AddressBlocklist) FilterAllowlist(allowlist common.AddressSet, latest *types.Header) common.AddressSet {
+	b.mux.RLock()
+	defer b.mux.RUnlock()
+
+	filtered := common.AddressSet{}
+	for a := range allowlist {
+		if !b.isBlocked(a, latest) {
+			filtered[a] = struct{}{}
+		}
+	}
+	return filtered
+}
+
+func (b *AddressBlocklist) IsBlocked(currency common.Address, latest *types.Header) bool {
+	b.mux.RLock()
```

Comment:
We also obtain a read lock in the `isBlocked` function. Is it intentional? Why not combine both or remove the lock from the `isBlocked` function like you did for `Evict`/`evict`? Or perhaps a `sync.Map` would be easier to use.

Reply:
Good catch, missed this one as the intention was exactly to only have the lock in the public method.

I'll look into the sync.Map, but I'm not sure if a cache with eviction is the correct use-case for this, as per the docs:

> The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys.

We do 1 write, tons of reads, and 1 delete - I don't know if this is still optimized for this